docs/fp_primer.md

```diff
@@ -0,0 +1,495 @@
+
+Additionally, floating point numbers are not evenly distributed over their in 
+interval, but instead are clustered closer together near zero, and further apart
+near the edges.
+
+When discussing operations on floating point numbers, there is often reference
+to a true value. This is the value that given no performance constraints and
+infinite precision you would get, i.e `acos(1) = π`, where π has infinite
+digits of precision.
+
+For the CTS it is often sufficient to calculate the true value using TypeScript,
+since its native number format is higher precision (double-precision/f64), and
+all f32 values can be represented in it.
+
+The true value is sometimes representable exactly as a f32 value, but often is
+not.
+
+Additionally, many operations are implemented using approximations from
+numerical analysis, where there is a tradeoff between the precision of the
+result and the cost.
+
+Thus, the spec specifies what the accuracy constraints for specific operations
+is, how close to truth an implementation is required to be, to be
+considered conformant.
+
+There are 5 different ways that accuracy requirements are defined in the spec:
+
+1. *Exact*
+
+   This is the situation where it is expected that true value for an operation
+   is always expected to be exactly representable. This doesn't happen for any
+   of the operations that return floating point values, but does occur for
+   logical operations that return boolean values.
+
+
+2. *Correctly Rounded*
+
+   For the case that the true value is exactly representable as a f32, this is
+   the equivalent of exactly from above. In the event that the true value is not
+   exact, then the acceptable answer for most numbers is the nearest f32 above
+   and below the true value. 
+
+   For values near the subnormal range, e.g. close to zero, this becomes more
+   complex, since an implementation may FTZ at any point. So if the exact 
+   solution is subnormal or either of the neighbours of the true value are
+   subnormal, zero becomes a possible result, thus the acceptance interval is 
+   wider than naively expected.
+
+
+3. *Absolute Error*
+   
+   This type of accuracy specifies an error value, ε, and the calculated result
+   is expected to be within that distance from the true value, i.e.
+   `[ X - ε, X + ε ]`.
+
+   The main drawback with this manner of specifying accuracy is that it doesn't
+   scale with the level of precision in floating point numbers themselves at a
+   specific value. Thus, it tends to be only used for specifying accuracy over
+   specific limited intervals, i.e. [-π, π].
+
+
+4. *Units of Least Precision (ULP)*
+
+   The solution to the issue of not scaling with precision of floating point is
+   to use units of least precision.
+
+   There are many more formal definitions of ULP, but at a high level it is the
+   difference from a specific value to the next nearest f32 value.
+
+   n * ULP or nULP means `[X - n * ULP @ X, X + n * ULP @ X]`.
+
```

Comment:
The geek in me begs for the formal definition of ULP here, after statement of n*ULP.  This is the place for being nerdly-correct.


> ULP(X) is   min (b-a) over all pairs (a,b) of representable floating point numbers such that (a <= X <= b)

Reply:
Done